web/pandas/pdeps/0009-io-extensions.md

```diff
@@ -0,0 +1,395 @@
+be allowed. By simply installing the appropriate packages users will
+be able to use code like this:
+
+```python
+import pandas
+
+df = pandas.read_duckdb("SELECT * FROM 'dataset.parquet';")
+
+df.to_hive(hive_conn, "hive_table")
+```
+
+This API allows for method chaining:
+
+```python
+(pandas.read_duckdb("SELECT * FROM 'dataset.parquet';")
+       .to_hive(hive_conn, "hive_table"))
+```
+
+The total number of I/O functions and methods is expected to be small, as users
+in general use only a small subset of formats. The number could actually be
+reduced from the current state if the less popular formats (such as SAS, SPSS,
+BigQuery, etc.) are removed from the pandas core into third-party packages.
+Moving these connectors is not part of this proposal, and will be discussed
+later in a separate proposal.
+
+#### Plugin registration
+
+Third-party packages would implement an
+[entrypoint](https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins)
+to define the connectors that they implement, under a group `dataframe.io`.
+
+For example, a hypothetical project `pandas_duckdb` implementing a `read_duckdb`
+function, could use `pyproject.toml` to define the next entry point:
+
+```toml
+[project.entry-points."dataframe.io"]
+from_duckdb = "pandas_duckdb:read_duckdb"
+```
+
+On import of the pandas module, it would read the entrypoint registry for the
+`dataframe.io` group, and would dynamically create methods in the `pandas` and
+`pandas.DataFrame` namespace for them. Method not starting with `read_` or `to_`
+would make pandas raise an exception. This would guarantee a reasonably
+consistent API among third-party connectors.
+
+#### Internal API
+
+Connectors would use one of two different interface options: a pandas `DataFrame`
+or an Apache Arrow table.
+
+The Apache Arrow format would allow that connectors do not need to use pandas to
+create the data, making them more robust and less likely to break when changes to
+pandas internals happen. It would also allow other possible consumers of the
+connectors to not have pandas as a dependency. Testing also becomes simpler by
+using Apache Arrow, since connectors can be tested independently, and pandas does
+not need to be tested for each connector. If the Apache Arrow specification is
+respected in both sides, the communication between connectors and pandas is
+guaranteed to work. If pandas eventually has a hard dependency on an Apache
+Arrow implementation, this should be the preferred interface.
+
+Allowing connectors to use pandas dataframes directly makes users not have to
+depend on PyArrow for connectors that do not use an Apache Arrow object. It also
+helps move existing connectors to this new API, since they are using pandas
+dataframes as an exchange object now. It has the disadvantages stated in the
+previous paragraph, and a future proposal may be created to discuss deprecating
+pandas dataframes as a possible interface for connectors.
```

Comment:
I left this open as a possibility. As said in the other comment, Arrow seems more robust. I can delete this comment, but I thought it could be worth noting that this can be discussed in the future.